integration-libs/opf/cta/core/services/opf-dynamic-cta.service.ts

```diff
@@ -0,0 +1,251 @@
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = true;
+    return combineLatest({
+      cart: this.activeCartFacade.takeActive(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ cart, additionalData }) => {
+        return {
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          ctaProductItems: cart.entries?.map((entry) => {
+            return {
+              productId: entry.product?.code,
+              quantity: entry.quantity,
+            } as CtaProductItem;
+          }),
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  fillCtaRequestforProductPage(
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = false;
+    return combineLatest({
+      product: this.currentProductService.getProduct(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ product, additionalData }) => {
+        return {
+          ctaProductItems: [{ productId: product?.code, quantity: 1 }],
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  protected fillAdditionalData(): Observable<Array<KeyValuePair>> {
+    return combineLatest({
+      language: this.languageService.getActive(),
+      currency: this.currencyService.getActive(),
+    }).pipe(
+      take(1),
+      map(({ language, currency }) => {
+        return [
+          {
+            key: 'locale',
+            value: language,
+          },
+          {
+            key: 'currency',
+            value: currency,
+          },
+          {
+            key: 'scriptIdentifier',
+            value: this.getNewScriptIdentifier(),
+          },
+        ];
+      })
+    );
+  }
+
+  initiateEvents() {
+    if (!this.isOnsiteMessagingInit) {
+      this.listenScriptReadyEvent();
+
+      this.isCartPage && this.cartChangedListener();
+      !this.isCartPage && this.productChangedListener();
+
+      this.isOnsiteMessagingInit = true;
+    }
+  }
+
+  stopEvents() {
+    if (this.isOnsiteMessagingInit) {
+      this.subList.forEach((sub) => sub.unsubscribe());
+      this.subList = [];
+      this.globalFunctionsFacade.removeGlobalFunctions(
+        GlobalFunctionsDomain.GLOBAL
+      );
+      this.isOnsiteMessagingInit = false;
+    }
+  }
+
+  protected getNewScriptIdentifier(): string {
+    this.scriptIdentifiers.push(String(this.scriptIdentifiers.length + 1));
+    return String(
+      this.scriptIdentifiers[this.scriptIdentifiers.length - 1]
+    ).padStart(4, '0');
+  }
+
+  registerScriptReadyEvent() {
+    this.globalFunctionsFacade.registerGlobalFunctions({
+      paymentSessionId: '',
+      domain: GlobalFunctionsDomain.GLOBAL,
+    });
+  }
+
+  protected listenScriptReadyEvent() {
```

Comment:
[MEDIUM]
IMHO the method name is a bit misleading, especially while being the same as `opfCtaFacade.listenScriptReadyEvent()` , but those 2 methods doing 2 very different things - this one subscribes to a scriptReadt event and dispatches native CTA events; but the other method just returns an observable emitting the the scriptReady event.

1. I suggest to rename the other one - `opfCtaFacade.listenScriptReadyEvent()` to `opfCtaFacade.getScriptReadyEvent()` - because the method itself doens't setup any event _listener_, but just returns the observable, so you can _get_ the event observable.

2. I suggest to also align the naming of the local method `listenScriptReadyEvent()` with `cartChangedListener()` (and `productChangedListener()`). They do analogical things - they all subscribe to some Spartacus events and then emit native CTA events. What about renaming them to:
`addScriptReadyEventListener()`
`addCartChangedEventListener()`
`addProductChangedEventListeners()`

3. Then we could also align `initiateEvents()` and `stopEvents()` and renamme them to (respectively): `addEventListeners()` and `removeEventListeners()`.
Moreover, the property `this.isOnsiteMessagingInit` could be renamed to `this.eventListenersAdded`.